src/util/cilfacade.ml

```diff
@@ -79,6 +79,81 @@ let end_basic_blocks f =
   let thisVisitor = new allBBVisitor in
   visitCilFileSameGlobals thisVisitor f
 
+class loopUnrollingVisitor = object
+  inherit nopCilVisitor
+  method! vstmt s =
+    (* All the duplicated stmts need to be re-created or the sid will fail*)
+    let newsid st = { st with sid = st.sid } in
+    let rec newsid_stmt st = 
+      let mkb stml = mkBlock (List.map newsid_stmt stml) in
+      match st.skind with
+      | Instr _ -> newsid st
+      | If(e,b1,b2,l1,l2) -> mkStmt(If(e,(mkb b1.bstmts),(mkb b2.bstmts),l1,l2))
+      | Loop(bl,l1,l2,s1,s2) -> 
+        let create_stmt = mkStmt(Loop((mkb bl.bstmts),l1,l2,s1,s2)) in
+        create_stmt.labels <- st.labels;
+        create_stmt
+      | Block(bl) -> mkStmt(Block(mkb bl.bstmts))
+      | Switch(e,bl,stl,l1,l2) -> mkStmt(Switch(e,(mkb bl.bstmts),(List.map newsid_stmt stl),l1,l2))
+      | _ -> newsid st in
+    let mkBlock' b = mkBlock (List.map newsid_stmt b) in
+    let mk_stmt_from_stmt_list bl = mkStmt(Block(mkBlock' bl)) in
+    match s.skind with
+    | Loop(b, loc, _, _, _) ->
+      let get_unrolling_factor = GobConfig.get_int "exp.unrolling-factor" in
+      (* All unrollings will leave the original loop at the end. The label makes sure it's not unrolled twice.*)
+      let add_label_to_remainder_loop st_loop loc_loop = 
+        st_loop.labels <- [Label(Cil.freshLabel "remainder_loop", loc_loop, false)];
+        st_loop in
+      let is_remainder_loop loop_stmt_labels =
+        match loop_stmt_labels with
+        | [] -> false
+        | Label(lab,_,_)::tl ->
+          begin
+            try
+              match String.sub lab 0 14 with
+              | "remainder_loop" -> true
+              | _ -> false
+            with _-> false
+          end
+        | _ -> false in
+      (* Because this is executed before preparedCFG, there are still Breaks instead of Gotos.*)
+      (* We need to transform them so we can replicate the loop's body outside of the loop.*)
+      let break_stmt = mkStmt (Instr []) in
+      break_stmt.labels <- [Label(Cil.freshLabel "unroll_while_break",loc,false)] ;
+      let rec rm_breaks_st st = 
+        let rm_breaks_st_list stl = mkBlock (List.map rm_breaks_st stl) in
+        match st.skind with
+        | Break(l) -> mkStmt (Goto (ref break_stmt, loc))
+        | If(e,tb,fb,l1,l2) -> mkStmt (If (e,(rm_breaks_st_list tb.bstmts), (rm_breaks_st_list fb.bstmts),l1,l2))
+        | Block(bl) -> mkStmt (Block (rm_breaks_st_list bl.bstmts))
+        |Switch(e,bl,stl,l1,l2) -> mkStmt (Switch (e, (rm_breaks_st_list bl.bstmts), (List.map rm_breaks_st stl),l1, l2))
+        | _ -> st in
```

Comment:
Why not just run this visitor _after_ `preparedCFG` and not reinvent the wheel?
Also this doesn't handle `continue`, which would have to jump to the next unrolled loop iteration.